op-e2e/e2eutils/geth/wait.go

```diff
@@ -99,8 +99,47 @@ func WaitForBlock(number *big.Int, client *ethclient.Client, timeout time.Durati
 			}
 		case err := <-headSub.Err():
 			return nil, fmt.Errorf("error in head subscription: %w", err)
-		case <-timeoutCh:
-			return nil, errTimeout
+		case <-ctx.Done():
+			return nil, ctx.Err()
 		}
 	}
 }
+
+func WaitForBlockToBeFinalized(number *big.Int, client *ethclient.Client, timeout time.Duration) (*types.Block, error) {
+	return waitForBlockTag(number, client, timeout, rpc.FinalizedBlockNumber)
+}
+
+func WaitForBlockToBeSafe(number *big.Int, client *ethclient.Client, timeout time.Duration) (*types.Block, error) {
+	return waitForBlockTag(number, client, timeout, rpc.SafeBlockNumber)
+}
+
+// waitForBlockTag polls for a block number to reach the specified tag & then returns that block at the number.
+func waitForBlockTag(number *big.Int, client *ethclient.Client, timeout time.Duration, tag rpc.BlockNumber) (*types.Block, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	// Wait for it to be finalized. Poll every half second.
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+
+	tagBigInt := big.NewInt(tag.Int64())
+
+	for range ticker.C {
+		block, err := client.BlockByNumber(ctx, tagBigInt)
+		if err != nil {
+			return nil, err
+		}
+		if block.NumberU64() >= number.Uint64() {
+			return client.BlockByNumber(ctx, number)
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			// Continue polling
+		}
```

Comment:
I think you need to restore the old `for { select { ...` loop, otherwise you're still blocked by the ticker, in case the context cancels early.

Reply:
The thing is, the linter was failing, saying the infinite `for` loop was illegal - that's why I added this ticker. I'll see if I can add the `for` loop back in a way that complies with the linter.

Reply:
Okay, I think I have a version that obeys without blocking.